packages/arb-validator-core/message/transactionBatch.go

```diff
@@ -0,0 +1,313 @@
+		hashing.Bytes32(hashing.SoliditySHA3(data)),
+	)
+}
+
+func (m DeliveredTransactionBatch) getTransactions() []DeliveredTransaction {
+	txes := make([]DeliveredTransaction, 0, len(m.Tos))
+	dataOffset := uint32(0)
+	for i := range m.Tos {
+		data := m.Data[dataOffset : dataOffset+m.DataLengths[i]]
+		offchainHash := OffchainTxHash(
+			m.Chain,
+			m.Tos[i],
+			m.SequenceNums[i],
+			m.Values[i],
+			data,
+		)
+		messageHash := hashing.SoliditySHA3WithPrefix(offchainHash[:])
+		sig := m.Signatures[i*65 : (i+1)*65]
+		pubkey, err := crypto.SigToPub(messageHash.Bytes(), sig)
+		if err != nil {
+			log.Fatalln("Invalid sig", err)
+		}
+
+		fromAddress := common.NewAddressFromEth(crypto.PubkeyToAddress(*pubkey))
+
+		tx := Transaction{
+			Chain:       m.Chain,
+			To:          m.Tos[i],
+			From:        fromAddress,
+			SequenceNum: m.SequenceNums[i],
+			Value:       m.Values[i],
+			Data:        data,
+		}
+
+		txes = append(txes, DeliveredTransaction{
+			Transaction: tx,
+			BlockNum:    m.BlockNum,
+			Timestamp:   m.Timestamp,
+		})
+		dataOffset += m.DataLengths[i]
+	}
+	return txes
+}
+
+func (m TransactionBatch) String() string {
+	return fmt.Sprintf("TransactionBatch()")
+}
+
+func (m TransactionBatch) Equals(other Message) bool {
+	o, ok := other.(TransactionBatch)
+	if !ok {
+		return false
+	}
+	if m.Chain != o.Chain ||
+		len(m.Tos) != len(o.Tos) ||
+		len(m.SequenceNums) != len(o.SequenceNums) ||
+		len(m.Values) != len(o.Values) ||
+		len(m.Data) != len(o.Data) ||
+		len(m.Signatures) != len(o.Signatures) {
+		return false
+	}
+
+	for i := range m.Tos {
+		if m.Tos[i] != o.Tos[i] {
+			return false
+		}
+	}
+	for i := range m.Tos {
+		if m.SequenceNums[i].Cmp(o.SequenceNums[i]) != 0 {
+			return false
+		}
+	}
+	for i := range m.Tos {
+		if m.Values[i].Cmp(o.Values[i]) != 0 {
+			return false
+		}
+	}
+	if !bytes.Equal(m.Data, o.Data) {
+		return false
+	}
+	if !bytes.Equal(m.Signatures, o.Signatures) {
+		return false
+	}
+
+	return true
+}
+
+func (m TransactionBatch) Type() MessageType {
+	return TransactionBatchType
+}
+
+type DeliveredTransactionBatch struct {
+	TransactionBatch
+	BlockNum  *common.TimeBlocks
+	Timestamp *big.Int
+}
+
+func (m DeliveredTransactionBatch) Equals(other Message) bool {
+	o, ok := other.(DeliveredTransactionBatch)
+	if !ok {
+		return false
+	}
+	return m.TransactionBatch.Equals(o.TransactionBatch) &&
+		m.BlockNum.Cmp(o.BlockNum) == 0 &&
+		m.Timestamp.Cmp(o.Timestamp) == 0
+}
+
+func (m DeliveredTransactionBatch) DeliveredHeight() *common.TimeBlocks {
+	return m.BlockNum
+}
+
+func (m DeliveredTransactionBatch) DeliveredTimestamp() *big.Int {
+	return m.Timestamp
+}
+
+func (m DeliveredTransactionBatch) CommitmentHash() common.Hash {
+	addressTy, _ := abi.NewType("address", "", nil)
+	addressArrayTy, _ := abi.NewType("address[]", "", nil)
+	uint256ArrayTy, _ := abi.NewType("uint256[]", "", nil)
+	uint32ArrayTy, _ := abi.NewType("uint32[]", "", nil)
+	bytesTy, _ := abi.NewType("bytes", "", nil)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [string literal "" appears, at least, 5 times, create a named constant for it](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5428340)